src/lib/device/virtual_dev.cpp

```diff
@@ -489,7 +194,7 @@ BlkAllocStatus VirtualDev::do_alloc_blk(blk_count_t nblks, const blk_alloc_hints
 }
 
 BlkAllocStatus VirtualDev::alloc_blk_from_chunk(blk_count_t nblks, const blk_alloc_hints& hints,
-                                                std::vector< BlkId >& out_blkid, PhysicalDevChunk* chunk) {
+                                                std::vector< BlkId >& out_blkid, Chunk* chunk) {
```

Comment:
May be I didn't get your suggestion correctly, but what you have suggested is exactly the current behavior.

Reply:
The current behavior is 
1. `blk_allocator_mutable()->alloc` may return partial allocation,  not sure if any other caller like the partial result but `VirtualDev::alloc_blk_from_chunk` doesnt . Suggesting if we can pass in a flag to disable partial allocation, instead of caller release partial allocation and fail the request.
```
auto status = chunk->blk_allocator_mutable()->alloc(nblks, hints, chunk_blkid);
    if (status == BlkAllocStatus::PARTIAL) {
        // free partial result
        for (auto const b : chunk_blkid) {
            auto const ret = chunk->blk_allocator_mutable()->free_on_realtime(b);
            HS_REL_ASSERT(ret, "failed to free on realtime");
        }
        chunk->blk_allocator_mutable()->free(chunk_blkid);
        status = BlkAllocStatus::FAILED;
    } 
  ```
  2.  in `VirtualDev::alloc_contiguous_blk`,  it calls `alloc_blk`, however `alloc_blk` is likely to return in-contiguous range. The current behavior is ASSERT in    `VirtualDev::alloc_contiguous_blk` if blk is not contiguous.  Suggesting passing the contiguous flag at least to `alloc_blk` so it can retry on other chunk.  Or even better pass all the way down to the allocator::alloc().
  
  
 Sum up I am suggesting move upper layer logic into `alloc` for better efficiency .

Reply:
On point #1: I don't think we can move the hints to blk_allocator to not return partial result because the blk_allocator is extermely involved code base already and on an extremely rare scenario we don't want more code path, but handle the failure outside.

On point #2: we will revise as part of BlkId Issue [#151]. The revision would remove a separate call and unify into one call alloc_blk().

Reply:
thx for clarification， yes it makes sense